Breaking change doc - clarify when adding fields to structs is allowed (#29597)
dotnet/docs: docs/core/compatibility/index.md

   The return value of a method or the type of a property or field cannot be modified. For example, the signature of a method that returns an <xref:System.Object> cannot be changed to return a <xref:System.String>, or vice versa.

> - ❌ **DISALLOWED: Adding an instance field to a struct that has no nonpublic fields**

Reviewer: ```suggestion
- ❌ **DISALLOWED: Adding an instance field to a struct that has only public fields or no fields**
```

Reviewer: The double negative is a little confusing I think.

>   If a struct has only public fields or has no fields at all, callers can declare locals of that struct type without calling the struct's ctor or first initializing the local to _default(T)_, so long as all public fields are set on the struct before first use. Adding any new fields - public or nonpublic - to such a struct is a source breaking change for these callers, as the compiler will now require the additional fields to be initialized.

Reviewer: ```suggestion
  If a struct has only public fields or has no fields at all, callers can declare locals of that struct type without calling the struct's constructor or first initializing the local to `default(T)`, so long as all public fields are set on the struct before first use. Adding any new fields - public or nonpublic - to such a struct is a source breaking change for these callers, as the compiler will now require the additional fields to be initialized.
```

>   If a struct has only public fields or has no fields at all, callers can declare locals of that struct type without calling the struct's ctor or first initializing the local to _default(T)_, so long as all public fields are set on the struct before first use. Adding any new fields - public or nonpublic - to such a struct is a source breaking change for these callers, as the compiler will now require the additional fields to be initialized.

Reviewer 2: Why does it matter whether or not there were fields before? If there were fields, adding a new field is a breaking change, no?

```C#
struct Foo {
   int X;
   int Y;
   int Z; // new
}

static Foo CreateFoo()
{
   return new Foo { X = 1, Y = 2 }; // error: Z is not initialized
}
```

Reviewer: Your example is covered by the description because all the existing fields are public (has no nonpublic fields). So it's disallowed. If any of the existing fields were private, then the struct would have previously had to be initialized to `default(T)`, and that would apply to the new field too. I'll make a suggestion to change the wording from "Adding an instance field to a struct that has no nonpublic fields" to "Adding an instance field to a struct that has only public fields or no fields".

>   Additionally, adding any new fields - public or nonpublic - to such a struct is a binary breaking change to callers who have applied `[SkipLocalsInit]` to their code. Since the compiler wasn't aware of these fields at compile time, it could emit IL which does not fully initialize the struct, leading to the struct being created from uninitialized stack data.

Reviewer: ```suggestion
  Additionally, adding any new fields - public or nonpublic - to a struct with no fields or only public fields is a binary breaking change to callers who've applied `[SkipLocalsInit]` to their code. Since the compiler wasn't aware of these fields at compile time, it could emit IL that doesn't fully initialize the struct, leading to the struct being created from uninitialized stack data.
```
